### PR TITLE
Add response wrapping to list operations

### PR DESCRIPTION
**Pull request by jefferai**

### CODE REVIEW

**Comment by LLBennett** [comment]
LGTM

File: vault/request_handling.go [hunk 1]
```diff
@@ -357,6 +357,25 @@ func (c *Core) handleLoginRequest(req *logical.Request) (*logical.Response, *log
 }
 
 func (c *Core) wrapInCubbyhole(req *logical.Request, resp *logical.Response) (*logical.Response, error) {
+	// Before wrapping, obey special rules for listing: if no entries are
+	// found, 404. This prevents unwrapping only to find empty data.
+	if req.Operation == logical.ListOperation {
+		if resp == nil || len(resp.Data) == 0 {
+			return nil, logical.ErrUnsupportedPath
+		}
+		keysInt, ok := resp.Data["keys"]
```
Comment (vishalnayak) [inline vault/request_handling.go]: Can we rename this to `keysRaw`? `keysInt` is confusing.

Comment (jefferai) [inline vault/request_handling.go]: Done!

File: command/list.go [hunk 2]
```diff
@@ -60,6 +60,10 @@ func (c *ListCommand) Run(args []string) int {
 			"No value found at %s", path))
 		return 1
 	}
+	if secret.WrapInfo != nil {
```
Comment (vishalnayak) [inline command/list.go]: Can we add `secret.WrapInfo.TTL != 0` also here?
Like: https://github.com/hashicorp/vault/blob/master/vault/request_handling.go#L57

File: command/unwrap.go [hunk 3]
```diff
@@ -67,6 +67,15 @@ func (c *UnwrapCommand) Run(args []string) int {
 		return PrintRawField(c.Ui, secret, field)
 	}
 
+	// Check if the original was a list response and format as a list if so
+	if secret.Data != nil &&
```
Comment (vishalnayak) [inline command/unwrap.go]: For now, this seems to be the only easiest option we have. A unambiguous way to distinguish the List operation from the rest during unwrapping would have been nice.

**Comment by vishalnayak** [comment]
Very minor comments, otherwise LGTM!

**Comment by vishalnayak** [comment]
Unwrap test is failing, BTW. Just FYI.

File: command/list.go (hunk 2 above)
Comment (jefferai) [inline command/list.go]: Will do.

**Comment by jefferai** [comment]
I'll make sure that's fixed before merging.

**Pull request merged**